model/jpa/src/main/java/org/keycloak/models/jpa/JpaUserCredentialStore.java

```diff
@@ -106,31 +110,27 @@ CredentialModel toModel(CredentialEntity entity) {
     }
 
     @Override
-    public List<CredentialModel> getStoredCredentials(RealmModel realm, UserModel user) {
-        List<CredentialEntity> results = getStoredCredentialEntities(realm, user);
-
-        // list is ordered correctly by priority (lowest priority value first)
-        return results.stream().map(this::toModel).collect(Collectors.toList());
+    public Stream<CredentialModel> getStoredCredentialsStream(RealmModel realm, UserModel user) {
+        return this.getStoredCredentialEntities(realm, user).map(this::toModel);
     }
 
-    private List<CredentialEntity> getStoredCredentialEntities(RealmModel realm, UserModel user) {
+    private Stream<CredentialEntity> getStoredCredentialEntities(RealmModel realm, UserModel user) {
         UserEntity userEntity = em.getReference(UserEntity.class, user.getId());
         TypedQuery<CredentialEntity> query = em.createNamedQuery("credentialByUser", CredentialEntity.class)
                 .setParameter("user", userEntity);
-        return query.getResultList();
+        return closing(query.getResultStream().sorted(Comparator.comparingInt(CredentialEntity::getPriority)));
     }
 
     @Override
-    public List<CredentialModel> getStoredCredentialsByType(RealmModel realm, UserModel user, String type) {
-        return getStoredCredentials(realm, user).stream().filter(credential -> type.equals(credential.getType())).collect(Collectors.toList());
+    public Stream<CredentialModel> getStoredCredentialsByTypeStream(RealmModel realm, UserModel user, String type) {
+        return getStoredCredentialsStream(realm, user).filter(credential -> type.equals(credential.getType()));
```

Comment:
Rather use `Objects.equals` as it's null safe.

Reply:
Noted!